File: c++/src/xdevs/examples/efp/Gpt.cpp
/*
 * Gpt.cpp
 *
 *  Created on: 8 de may. de 2016
 *      Author: jlrisco
 */

#include "Gpt.h"

Gpt::Gpt(const std::string& name, const double& period, const double& observationTime) : Coupled(name), generator("generator", period),	processor("processor", 3*period), transducer("transducer", observationTime) {
	Coupled::addComponent(&generator);
	Coupled::addComponent(&processor);
	Coupled::addComponent(&transducer);
	Coupled::addCoupling(&generator, &generator.oOut, &processor, &processor.iIn);
	Coupled::addCoupling(&generator, &generator.oOut, &transducer, &transducer.iArrived);
	Coupled::addCoupling(&processor, &processor.oOut, &transducer, &transducer.iSolved);
	Coupled::addCoupling(&transducer, &transducer.oOut, &generator, &generator.iStop);
}

Gpt::~Gpt() { }

File: src/xdevs/core/examples/efp/Gpt.java
/*
 * Copyright (C) 2014-2015 José Luis Risco Martín <jlrisco@ucm.es> and 
 * Saurabh Mittal <smittal@duniptech.com>.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see
 * http://www.gnu.org/licenses/
 *
 * Contributors:
 *  - José Luis Risco Martín
 */
package xdevs.core.examples.efp;

import java.util.logging.Level;

import xdevs.core.modeling.Coupled;
import xdevs.core.simulation.Coordinator;
import xdevs.core.util.DevsLogger;

/**
 *
 * @author jlrisco
 */
public class Gpt extends Coupled {

    public Gpt(String name, double period, double observationTime) {
    	super(name);
        Generator generator = new Generator("generator", period);
        super.addComponent(generator);
        Processor processor = new Processor("processor", 3*period);
        super.addComponent(processor);
        Transducer transducer = new Transducer("transducer", observationTime);
        super.addComponent(transducer);
        
        super.addCoupling(generator.oOut, processor.iIn);
        super.addCoupling(generator.oOut, transducer.iArrived);
        super.addCoupling(processor.oOut, transducer.iSolved);
        super.addCoupling(transducer.oOut, generator.iStop);
    }

    public static void main(String args[]) {
        DevsLogger.setup(Level.FINE);
        Gpt gpt = new Gpt("gpt", 1, 100);
        //CoordinatorParallel coordinator = new CoordinatorParallel(gpt);
        Coordinator coordinator = new Coordinator(gpt);
        //RTCentralCoordinator coordinator = new RTCentralCoordinator(gpt);
        coordinator.initialize();
        coordinator.simulate(Long.MAX_VALUE);
    }

}
